File: engine/__tests__/unit/parser-pass.test.ts
import { describe, it, expect } from "vitest";
import { parse } from "../../passes/parser-pass.js";
import type { ASTNode } from "../../types.js";

const char = (v: string): ASTNode => ({ kind: "char", value: v });
const concat = (l: ASTNode, r: ASTNode): ASTNode => ({ kind: "concat", left: l, right: r });
const alt = (l: ASTNode, r: ASTNode): ASTNode => ({ kind: "alt", left: l, right: r });
const star = (c: ASTNode): ASTNode => ({ kind: "star", child: c });

describe("ParserPass — valid patterns (golden AST shapes)", () => {
  it("single char", () => {
    const r = parse("a");
    expect(r).toEqual({ ok: true, ast: char("a") });
  });

  it("concatenation is left-associative", () => {
    const r = parse("abc");
    expect(r).toEqual({ ok: true, ast: concat(concat(char("a"), char("b")), char("c")) });
  });

  it("alternation is left-associative", () => {
    const r = parse("a|b|c");
    expect(r).toEqual({ ok: true, ast: alt(alt(char("a"), char("b")), char("c")) });
  });

  it("star binds tighter than concatenation", () => {
    // "ab*" means a, then (b)* — NOT (ab)*
    const r = parse("ab*");
    expect(r).toEqual({ ok: true, ast: concat(char("a"), star(char("b"))) });
  });

  it("concatenation binds tighter than alternation", () => {
    // "ab|c" means (ab)|c — NOT a(b|c)
    const r = parse("ab|c");
    expect(r).toEqual({ ok: true, ast: alt(concat(char("a"), char("b")), char("c")) });
  });

  it("parentheses override precedence", () => {
    const r = parse("(a|b)*");
    expect(r).toEqual({ ok: true, ast: star(alt(char("a"), char("b"))) });
  });

  it("nested groups", () => {
    const r = parse("((a))");
    expect(r).toEqual({ ok: true, ast: char("a") });
  });

  it("backslash-escapes a metacharacter into a literal char", () => {
    const r = parse("\\(");
    expect(r).toEqual({ ok: true, ast: char("(") });
  });
});

describe("ParserPass — malformed patterns (golden ParseError shapes)", () => {
  it("empty pattern", () => {
    const r = parse("");
    expect(r).toEqual({
      ok: false,
      error: { type: "UnexpectedToken", position: 0, expected: "a character or '('", found: "end of input" },
    });
  });

  it("unmatched opening paren", () => {
    const r = parse("(a");
    expect(r).toEqual({
      ok: false,
      error: { type: "UnmatchedParen", position: 0, expected: "')'", found: "end of input" },
    });
  });

  it("unmatched closing paren", () => {
    const r = parse("a)");
    expect(r).toEqual({
      ok: false,
      error: { type: "UnmatchedParen", position: 1, expected: "end of input", found: ")" },
    });
  });

  it("empty group", () => {
    const r = parse("()");
    expect(r).toEqual({
      ok: false,
      error: { type: "EmptyGroup", position: 1, expected: "a pattern before ')'", found: ")" },
    });
  });

  it("leading star has nothing to repeat", () => {
    const r = parse("*a");
    expect(r).toEqual({
      ok: false,
      error: { type: "DanglingOperator", position: 0, expected: "a character or '('", found: "*" },
    });
  });

  it("double alternation operator", () => {
    const r = parse("a||b");
    expect(r).toEqual({
      ok: false,
      error: { type: "DanglingOperator", position: 2, expected: "a character or '('", found: "|" },
    });
  });

  it("trailing backslash with nothing to escape", () => {
    const r = parse("a\\");
    expect(r).toEqual({
      ok: false,
      error: { type: "UnexpectedToken", position: 1, expected: "a character after '\\'", found: "end of input" },
    });
  });
});

File: tests/parser_pass_test.cpp
// tests/parser_pass_test.cpp
//
// Direct port of parser-pass.test.ts. Same patterns, same golden AST/error
// shapes — TS's `toEqual` is replaced here by the operator== defined in
// types.hpp specifically for this purpose.

#include <catch2/catch_test_macros.hpp>

#include "regexlab/passes/parser-pass.hpp"
#include "regexlab/types.hpp"

using namespace regexlab;

namespace {

ParseResult ok(ASTNodePtr ast) {
    return ParseResult{true, ast, ParseError{}};
}
ParseResult err(ParseErrorType type, int position, std::optional<std::string> expected,
                std::optional<std::string> found) {
    return ParseResult{false, nullptr, ParseError{type, position, expected, found}};
}

}  // namespace

TEST_CASE("ParserPass — valid patterns (golden AST shapes)", "[parser]") {
    SECTION("single char") {
        REQUIRE(parse("a") == ok(makeChar('a')));
    }

    SECTION("concatenation is left-associative") {
        REQUIRE(parse("abc") ==
                ok(makeConcat(makeConcat(makeChar('a'), makeChar('b')), makeChar('c'))));
    }

    SECTION("alternation is left-associative") {
        REQUIRE(parse("a|b|c") ==
                ok(makeAlt(makeAlt(makeChar('a'), makeChar('b')), makeChar('c'))));
    }

    SECTION("star binds tighter than concatenation") {
        // "ab*" means a, then (b)* — NOT (ab)*
        REQUIRE(parse("ab*") == ok(makeConcat(makeChar('a'), makeStar(makeChar('b')))));
    }

    SECTION("concatenation binds tighter than alternation") {
        // "ab|c" means (ab)|c — NOT a(b|c)
        REQUIRE(parse("ab|c") == ok(makeAlt(makeConcat(makeChar('a'), makeChar('b')), makeChar('c'))));
    }

    SECTION("parentheses override precedence") {
        REQUIRE(parse("(a|b)*") == ok(makeStar(makeAlt(makeChar('a'), makeChar('b')))));
    }

    SECTION("nested groups") {
        REQUIRE(parse("((a))") == ok(makeChar('a')));
    }

    SECTION("backslash-escapes a metacharacter into a literal char") {
        REQUIRE(parse("\\(") == ok(makeChar('(')));
    }
}

TEST_CASE("ParserPass — malformed patterns (golden ParseError shapes)", "[parser]") {
    SECTION("empty pattern") {
        REQUIRE(parse("") == err(ParseErrorType::UnexpectedToken, 0,
                                  std::string("a character or '('"), std::string("end of input")));
    }

    SECTION("unmatched opening paren") {
        REQUIRE(parse("(a") == err(ParseErrorType::UnmatchedParen, 0, std::string("')'"),
                                    std::string("end of input")));
    }

    SECTION("unmatched closing paren") {
        REQUIRE(parse("a)") == err(ParseErrorType::UnmatchedParen, 1, std::string("end of input"),
                                    std::string(")")));
    }

    SECTION("empty group") {
        REQUIRE(parse("()") == err(ParseErrorType::EmptyGroup, 1, std::string("a pattern before ')'"),
                                    std::string(")")));
    }

    SECTION("leading star has nothing to repeat") {
        REQUIRE(parse("*a") == err(ParseErrorType::DanglingOperator, 0,
                                    std::string("a character or '('"), std::string("*")));
    }

    SECTION("double alternation operator") {
        REQUIRE(parse("a||b") == err(ParseErrorType::DanglingOperator, 2,
                                      std::string("a character or '('"), std::string("|")));
    }

    SECTION("trailing backslash with nothing to escape") {
        REQUIRE(parse("a\\") == err(ParseErrorType::UnexpectedToken, 1,
                                     std::string("a character after '\\'"), std::string("end of input")));
    }
}
